### ISSUE TITLE
Numerical differences on Python 3

### ISSUE THREAD

**Issue by jehturner** [iraf-community/pyraf#127]
After the recent work to get all our Gemini IRAF tests running, I've had a "quick" look at the differences in results compared with PyRAF 2.1.15 on Python 2.7, suspecting that most of the 189/932 failures may be caused by just a couple of small issues, which does indeed seem to be the case.

The first thing is that `iraffunctions.nint()` follows [the new behaviour of `round()` in Python 3](https://docs.python.org/3/whatsnew/3.0.html#builtins), rounding half-integers to the nearest even number instead of up. So eg. CL and PyRAF on Python 2 both give `nint(2.5) == 3`, whereas PyRAF on Python 3 gives `2`.

The other, [less well-documented](https://stackoverflow.com/a/25899600/1896081) thing is that Python 3 `float.__str__()` produces 17-digit precision (like `repr`), rather than 12 digits in Python 2. Part of the issue there is just our tests being overly sensitive/brittle. However, it can lead to some odd consequences, such as comments getting truncated where there's less space or `imexpr` switching to double precision output when given extra digits. It can also produce somewhat larger differences where the outcome of pixel rejection differs here and there, as well as uglier user output in some cases. 

So which is the "correct" behaviour? I think the baseline intention here is to emulate CL. For example, division has already been made to follow the CL & Python 2 behaviour on Python 3: `1/2 == 0`. I think `nint()` is analogous to that and a lot of CL code has been written assuming that it always rounds upwards. At a glance, I can see some cases where breaking that assumption causes arcane bugs, eg. a large imaging WCS error due to the WCS getting copied from the wrong input FITS extension, because the reference extension number is now getting calculated incorrectly. "Fixing" `nint()` is a 2-line patch that I think would be worth including, for compatibility with CL (including the ability to write scripts that work with both `cl` and `pyraf`) and to avoid unnecessary work.

The argument for trying to emulate the precision of CL's `real()` is less obvious though. First, it's a bit of an uphill struggle to override a built-in type, with Python's native precision likely to crop up wherever any function fails to override it. Second, it seems to be coincidental that Python 2's `float()` behaves so much like IRAF's `real()`, printing 12 significant digits, switching to scientific notation at the same threshold etc. I do think IRAF has occasionally influenced Python and it may be that their considerations were similar here, but I don't see any evidence of PyRAF _deliberately_ taking measures to reproduce the format shown by CL on Python 2; in fact, I think I found some corner case where it differs slightly, when looking into it. Third, an increase in precision should be numerically harmless. Fourth, there are other factors leading to (larger) floating-point differences anyway, such as what C library version IRAF was linked against. Fifth, CL scripts could explicitly format numeric output if they really care. So maybe it's better to accept this one and edit our tests and code to cope with it; I still have to finish looking into that.

Having said all of the above(!), since I had 189 failures to characterize, I actually _did_ have a quick go at hacking PyRAF to reproduce Python 2's `float.__str__()` behaviour, successfully. Combining those changes with the `nint()` fix and a few more fixes in our tests (and disregarding 14 tests that still fail for reasons unrelated to Python 3), I was able to prove that all of the remaining failures are due to those language changes in Python 3, which I think is a useful point of reference and means that we're close to being able to say that Gemini IRAF works on Python 3.

**Comment by jehturner** [comment]
These were the changes I made in `stsci.tools` & `pyraf` to emulate Python 2 `float()` for testing purposes, FYI:

https://github.com/jehturner/stsci.tools/commit/29363f5970d53b21b8754933455ddb4414d30365
https://github.com/jehturner/pyraf/commit/564b762eed5c988ceee644f4a87cc5c06d755e52

(Actually, I was testing a very slightly older pyraf revision -- these are the same changes rebased on the latest main branch and my `nint()` fix; I still have to test this copy, but you can see what it does.)

**Comment by jehturner** [comment]
It turns out that the above `stsci.tools` patch (along with the `nint()` fix) is enough to make all the tests pass by itself, without needing the PyRAF `float` patches that I did first. I suppose that's because all text I/O (except on the terminal) happens via parameters, so that's a well-defined point at which the format can be controlled. The internal `float` type within a CL script doesn't really matter, since it only affects conversion to `str` and not the precision of calculations per se. So that would make for a smaller "fix" that actually doesn't have much scope for corner cases to slip through the net after all -- but whether you (& STScI) really want to emulate CL & Python 2 in passing 12-digit precision instead of 17 is another matter. If not, that leaves us about 115 tests to fix (some of which will need new reference files because of issues like pixel rejection with data quality or awkward text comparison, which makes it difficult to support Python 2 & 3 at the same time until we can finish switching over). Let me know what you think...

**Comment by olebole** [comment]
@jehturner sorry, I am quite busy with other things in the moment. I'll have a look in a few days.

**Comment by jehturner** [comment]
OK, thanks (that's helpful to know, so I can go ahead with the next update using Python 2). BTW, I think this will be the last issue from Python 3 testing.

**Comment by jehturner** [comment]
I'll copy @pllim here, since one option would be to patch `stsci.tools` for stricter CL compatibility (on Python 3) as per the draft at https://github.com/jehturner/stsci.tools/commit/29363f5970d53b21b8754933455ddb4414d30365. But we should also wait for Ole's input, since it's PyRAF that's affected.

**Comment by pllim** [comment]
Let's wait. STScI does not officially support PyRAF anymore, so patching `stsci.tools` does not quite make sense if there are other options.

**Comment by pllim** [comment]
Maaaaybe TEAL... @cdsontag or @stsci-hack might know.

**Comment by pllim** [comment]
> Mihai would know (could not find his tag)

@mcara

**Comment by mcara** [comment]
not me. cc: @stsci-hack

**Comment by olebole** [comment]
[stsci.tools](https://github.com/spacetelescope/stsci.tools) looks to me quite PyRAF specific in most functions, so I would agree that we divorce here. With #129, we save all required code from there, and remove it as a dependency.

When it comes to the patch https://github.com/jehturner/stsci.tools/commit/29363f5970d53b21b8754933455ddb4414d30365 - I am nor really sure. Overwriting float as a base class seems a bit dangerous to me (although I can't specify a danger here), and doing this just to remove some digits of the output may be introducing a complex change for a small issue. Usually, tests shouldn't fail just because of minimal changes in floating point numbers.
But, on the other hand, it may not matter too much here -- PyRAF is an old package, and keeping these tests running without changes is worth it. So, @jehturner if you would create a PR here (after merging #129), I would approve it pragmatically.

**Comment by jehturner** [comment]
I think it would be rare for anyone to use more than 12 digits of precision in IRAF, so in that sense it doesn't really matter.

**Comment by jehturner** [comment]
Just some example differences:
```
> lpar gscalibrate
        input =                 Input spectra to calibrate
      (output = "")             Output calibrated spectra
     (outpref = "c")            Output prefix
   (sfunction = "sens")         Input image root name for sensitivity function
     (sci_ext = "SCI")          Name of science extension
     (var_ext = "VAR")          Name of variance extension
      (dq_ext = "DQ")           Name of data quality extension
 (key_airmass = "AIRMASS")      Airmass header keyword
 (key_exptime = "EXPTIME")      Exposure time header keyword
    (fl_vardq = no)             Propagate VAR/DQ planes
      (fl_ext = no)             Apply extinction correction to input spectra
     (fl_flux = yes)            Apply flux calibration to input spectra
    (fl_scale = yes)            Multiply output with fluxscale
   (fluxscale = 1000000000000000.0) Value of the flux scale (fl_scale=yes)
   (ignoreaps = yes)            Ignore aperture numbers in flux calibration
      (fl_fnu = no)             Create spectra having units of FNU
  (extinction = "")             Extinction file
 (observatory = "Gemini-North") Observatory
     (logfile = "")             Logfile name
     (verbose = yes)            Verbose?
      (status = 0)              Exit status (0=good)
    (scanfile = )               For internal use only
        (mode = "al")           
```
Versus this in Python 2 (or with the patch) -- note `fluxscale`:
```
        input =                 Input spectra to calibrate
      (output = "")             Output calibrated spectra
     (outpref = "c")            Output prefix
   (sfunction = "sens")         Input image root name for sensitivity function
     (sci_ext = "SCI")          Name of science extension
     (var_ext = "VAR")          Name of variance extension
      (dq_ext = "DQ")           Name of data quality extension
 (key_airmass = "AIRMASS")      Airmass header keyword
 (key_exptime = "EXPTIME")      Exposure time header keyword
    (fl_vardq = no)             Propagate VAR/DQ planes
      (fl_ext = no)             Apply extinction correction to input spectra
     (fl_flux = yes)            Apply flux calibration to input spectra
    (fl_scale = yes)            Multiply output with fluxscale
   (fluxscale = 1e+15)          Value of the flux scale (fl_scale=yes)
   (ignoreaps = yes)            Ignore aperture numbers in flux calibration
      (fl_fnu = no)             Create spectra having units of FNU
  (extinction = "")             Extinction file
 (observatory = "Gemini-North") Observatory
     (logfile = "")             Logfile name
     (verbose = yes)            Verbose?
      (status = 0)              Exit status (0=good)
    (scanfile = )               For internal use only
        (mode = "al")           
```
(That one actually _does_ result in numerical differences in the output, because the extra digits cause `imexpr` to use double precision, but my main point here is just to show you how Python 3 affects the UI.)

Headers:
```
CD1_1   =   -4.16666666667e-05 / WCS matrix element 1 1
CD1_2   =    1.11645402438e-12 / WCS matrix element 1 2
CD2_1   =   -1.11645402438e-12 / WCS matrix element 2 1
CD2_2   =   -4.16666666667e-05 / WCS matrix element 2 2
```
in Python 2 becomes this in Python 3 (one side effect is that the comment gets pushed past the usual column, which apparently breaks an assumption somewhere in our test comparisons):
```
CD1_1   =-4.166666666666665e-05 / WCS matrix element 1 1
CD1_2   =1.116454024376193e-12 / WCS matrix element 1 2
CD2_1   =-1.116454024376193e-12 / WCS matrix element 2 1
CD2_2   =-4.166666666666665e-05 / WCS matrix element 2 2
```
And cases where the terminal output looks like eg. `148948.80000000002` instead of `148948.8` or `0.9412735178174999` instead of `0.941273517817`. Plus differences in pixel values that include some of the more difficult cases, but I don't really need to show you those to get the idea :wink:.

**Comment by olebole** [comment]
Yea, can you make a PR with your change? I'd then apply it.

**Comment by jehturner** [comment]
OK then, thanks!

**Comment by jehturner** [comment]
Just to be clear, `pyraf` now completely independent of `stsci.tools`, right?

**Comment by olebole** [comment]
True. I actually forgot three files, which are however now in the main branch.

**Comment by jehturner** [comment]
The tests no longer work for me after the `stsci.tools` merge; I get a big traceback (apparently as of `63c53260`), which I haven't had time to troubleshoot properly today. Does this work for you? Thanks.
```
> pytest -s pyraf/tests/test_cli.py
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.7.9, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
DRAGONS_TEST directory: None
rootdir: /home/jturner/repos/pyraf, configfile: setup.cfg
plugins: hypothesis-6.8.5, arraydiff-0.3, astropy-header-0.1.2, doctestplus-0.0.0, openfiles-0.5.0, remotedata-0.3.2, asdf-2.8.0.dev224+gba09e57.d20210407, dragons-3.1.0.dev0
collected 0 items                                                                                                                                                                                                 

============================================================================================== no tests ran in 0.12s ==============================================================================================
Traceback (most recent call last):
  File "/home/jturner/repos/pyraf/pyraf/irafpar.py", line 164, in makeIrafPar
    return IrafParFactory(fields, strict=strict)
  File "/home/jturner/repos/pyraf/pyraf/irafpar.py", line 52, in IrafParFactory
    return basicpar.parFactory(fields, strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 68, in parFactory
    return IrafParI(fields,strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 171, in __init__
    self.value = self._coerceValue(fields[3],strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 759, in _coerceValue
    return self._coerceOneValue(value,strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 1362, in _coerceOneValue
    (self.name, repr(value)))
ValueError: Parameter $nargs: illegal integer value '<IrafParI $nargs i h 0 None None "">'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/main.py", line 257, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/main.py", line 312, in _main
    config.hook.pytest_collection(session=session)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/main.py", line 323, in pytest_collection
    session.perform_collect()
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/main.py", line 626, in perform_collect
    self.items.extend(self.genitems(node))
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/main.py", line 800, in genitems
    rep = collect_one_node(node)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/runner.py", line 457, in collect_one_node
    rep = ihook.pytest_make_collect_report(collector=collector)  # type: CollectReport
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/capture.py", line 803, in pytest_make_collect_report
    rep = outcome.get_result()
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/runner.py", line 359, in pytest_make_collect_report
    errorinfo = collector.repr_failure(call.excinfo)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/nodes.py", line 485, in repr_failure
    return self._repr_failure_py(excinfo, style=tbstyle)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/nodes.py", line 415, in _repr_failure_py
    truncate_locals=truncate_locals,
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 627, in getrepr
    return fmt.repr_excinfo(self)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 884, in repr_excinfo
    reprtraceback = self.repr_traceback(excinfo_)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 825, in repr_traceback
    reprentry = self.repr_traceback_entry(entry, einfo)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 777, in repr_traceback_entry
    s = self.get_source(source, line_index, excinfo, short=short)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 715, in get_source
    lines.extend(self.get_exconly(excinfo, indent=indent, markall=True))
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 727, in get_exconly
    exlines = excinfo.exconly(tryshort=True).split("\n")
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 546, in exconly
    lines = format_exception_only(self.type, self.value)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 140, in format_exception_only
    return list(TracebackException(etype, value, None).format_exception_only())
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 497, in __init__
    _seen=_seen)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 508, in __init__
    capture_locals=capture_locals)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 359, in extract
    linecache.checkcache(filename)
  File "/home/jturner/repos/pyraf/pyraf/cllinecache.py", line 30, in checkcache
    from . import iraf  # used below
  File "/home/jturner/repos/pyraf/pyraf/irafimport.py", line 87, in _irafImport
    retval = _originalImport(name, globals, locals, fromlist, level)
  File "/home/jturner/repos/pyraf/pyraf/__init__.py", line 68, in <module>
    iraf.Init(doprint=0, hush=1)
  File "/home/jturner/repos/pyraf/pyraf/iraffunctions.py", line 208, in Init
    clProcedure(Stdin='hlib$zzsetenv.def')
  File "/home/jturner/repos/pyraf/pyraf/iraffunctions.py", line 2749, in clProcedure
    newtask.run()
  File "/home/jturner/repos/pyraf/pyraf/iraftask.py", line 373, in run
    self._run(redirKW, specialKW)
  File "/home/jturner/repos/pyraf/pyraf/iraftask.py", line 1521, in _run
    self._runCode()
  File "/home/jturner/repos/pyraf/pyraf/iraftask.py", line 1537, in _runCode
    self._clFunction(*parList, **kw)
  File "<CL script /home/jturner/anaconda3_2019>", line 10, in zzsetenv
  File "/home/jturner/repos/pyraf/pyraf/irafpar.py", line 167, in makeIrafPar
    raise ValueError(errmsg)
ValueError: Bad value for parameter `$nargs'
Parameter $nargs: illegal integer value '<IrafParI $nargs i h 0 None None "">'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jturner/repos/pyraf/pyraf/irafpar.py", line 164, in makeIrafPar
    return IrafParFactory(fields, strict=strict)
  File "/home/jturner/repos/pyraf/pyraf/irafpar.py", line 52, in IrafParFactory
    return basicpar.parFactory(fields, strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 68, in parFactory
    return IrafParI(fields,strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 171, in __init__
    self.value = self._coerceValue(fields[3],strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 759, in _coerceValue
    return self._coerceOneValue(value,strict)
  File "/home/jturner/repos/pyraf/pyraf/tools/basicpar.py", line 1362, in _coerceOneValue
    (self.name, repr(value)))
ValueError: Parameter $nargs: illegal integer value '<IrafParI $nargs i h 0 None None "">'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jturner/anaconda3_2019.10/envs/spec37/bin/pytest", line 33, in <module>
    sys.exit(load_entry_point('pytest', 'console_scripts', 'pytest')())
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/config/__init__.py", line 187, in console_main
    code = main()
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/config/__init__.py", line 165, in main
    config=config
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/main.py", line 306, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/main.py", line 279, in wrap_session
    config.notify_exception(excinfo, config.option)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/config/__init__.py", line 1035, in notify_exception
    funcargs=True, showlocals=getattr(option, "showlocals", False), style=style
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/site-packages/_pytest/_code/code.py", line 612, in getrepr
    self.type, self.value, self.traceback[0]._rawentry
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 121, in format_exception
    type(value), value, tb, limit=limit).format(chain=chain))
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 497, in __init__
    _seen=_seen)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 508, in __init__
    capture_locals=capture_locals)
  File "/home/jturner/anaconda3_2019.10/envs/spec37/lib/python3.7/traceback.py", line 359, in extract
    linecache.checkcache(filename)
  File "/home/jturner/repos/pyraf/pyraf/cllinecache.py", line 30, in checkcache
    from . import iraf  # used below
  File "/home/jturner/repos/pyraf/pyraf/irafimport.py", line 87, in _irafImport
    retval = _originalImport(name, globals, locals, fromlist, level)
  File "/home/jturner/repos/pyraf/pyraf/__init__.py", line 68, in <module>
    iraf.Init(doprint=0, hush=1)
  File "/home/jturner/repos/pyraf/pyraf/iraffunctions.py", line 208, in Init
    clProcedure(Stdin='hlib$zzsetenv.def')
  File "/home/jturner/repos/pyraf/pyraf/iraffunctions.py", line 2749, in clProcedure
    newtask.run()
  File "/home/jturner/repos/pyraf/pyraf/iraftask.py", line 373, in run
    self._run(redirKW, specialKW)
  File "/home/jturner/repos/pyraf/pyraf/iraftask.py", line 1521, in _run
    self._runCode()
  File "/home/jturner/repos/pyraf/pyraf/iraftask.py", line 1537, in _runCode
    self._clFunction(*parList, **kw)
  File "<CL script /home/jturner/anaconda3_2019>", line 10, in zzsetenv
  File "/home/jturner/repos/pyraf/pyraf/irafpar.py", line 167, in makeIrafPar
    raise ValueError(errmsg)
ValueError: Bad value for parameter `$nargs'
Parameter $nargs: illegal integer value '<IrafParI $nargs i h 0 None None "">'
```

**Comment by olebole** [comment]
The integration requires to re-populate the cache, since the code generated with cl2py had an stsci.tools import statement https://github.com/iraf-community/pyraf/blob/32455a4a4e1085b17b5a31ab767e4dd6b74dcbac/pyraf/cl2py.py#L1676

This is replaced with a pyraf.tools import statement: https://github.com/iraf-community/pyraf/blob/397247842e68f6fce5ba90c8301be005b39c8399/pyraf/cl2py.py#L1672

The stsci.tools import in cached files interferes with the pyraf.tools one in the updated code. In principle, I could introduce a new cache file version "v3" or so; maybe this is a good idea?

**Comment by jehturner** [comment]
I've noted in #133 that enforcing Python 2 `str(float)` precision doesn't actually mimic CL in some cases. Without that patch, however, it turns out that some of the headers produced are not just slightly different on Python 3 -- they actually violate the FITS standard (and cause warnings from AstroPy). The CD matrix in https://github.com/iraf-community/pyraf/issues/127#issuecomment-1064204628 above illustrates that. The FITS standard (S4.1.2) requires a space in column 10, after the `=`, but here the floating-point value is spilling into that column (this happens most often when there's a minus sign). Also, while the standard does allow for longer "free-format" values, I believe these are intended to be "fixed-format" floating point numbers that should end in column 30 (S4.2.4), whereas in the above example, the exponent spills into column 31. On Python 2, the shorter numbers get right-justified at column 30, as per S4.2.4, with several spaces after the `=`. I haven't identified the code responsible for this yet, but it seems that if we were to decide against merging #133 after all, some other bit of code would need fixing to resolve this.

**Comment by jehturner** [comment]
This turns out to be an IRAF bug that affects both `nhedit` and `hedit` when used to update en existing floating-point header value -- eg.:
```
> imcopy dev$pix test.fits
> hedit("test.fits", "CD1_1", -1.2345678901234567E-08, add=yes, update=yes, verify=no)
> imhead test.fits l+ |match CD1
CD1_1   =         -1.234568E-8
> hedit("test.fits", "CD1_1", -1.2345678901234567E-08, add=yes, update=yes, verify=no)
> imhead test.fits l+ |match CD1
CD1_1   =-1.2345678901234567e-08
```
Note that the `hedit` command is identical in both cases -- the only difference is that the floating-point header entry already exists the second time.

This does _not_ happen in CL when following the above example, because (like Python 2) its numerical precision is lower -- but you can make it happen in CL by converting the value to a string in the second `hedit` call -- in that case, the header entry retains its existing floating-point type, but using a string causes the extra digits to get passed to `hedit` (as with PyRAF). So strictly it's a core IRAF bug. In the real World, the telescope/instrument will write a FITS header with a floating-point CD matrix and then the problem will occur when the user does `hedit` using a string value (or with PyRAF).

I don't know whether community IRAF also has this bug, but obviously restricting the precision in PyRAF is one way of avoiding it... Another would be to fix both our IRAF distribution and yours (but it's more of a corner case in CL than PyRAF) and a third would be to restrict the precision passed to `nhedit` by Gemini IRAF, but that's not a general solution.

**Comment by jehturner** [comment]
From a quick look at the code, I suspect the culprits are `iraf/sys/imio/db/impstr.x` and `iraf/sys/imio/dbc/impstrc.x`.

**Comment by jehturner** [comment]
Incidentally, I've noticed a change in the behaviour of PyRAF's tab completion. When I type the name of a directory and press tab, I now get tab characters after the trailing slash, instead of a directory listing. Is that something you already know about or should I make an issue for it? I'm not sure when it started happening but it just began annoying me today :slightly_smiling_face:.

**Comment by olebole** [comment]
I observe the same (and didn't see this before), please open an issue for it.

**Comment by olebole** [comment]
This is closed by #133
